### DISCUSSION TITLE
Documentation on how to use this library using AWS lambda

### DISCUSSION THREAD

**Discussion by kishaningithub** [aws/amazon-redshift-python-driver#78] [Ideas] [answered]
When i was trying to use this using lambda, i had to remove following deps from the zip. Have put in the reasons for the same

- botocore 
   - Part of AWS Lambda runtime (https://www.feitsui.com/en/article/2)
- boto3
   - Part of AWS Lambda runtime (https://www.feitsui.com/en/article/2)
- lxml
   -  Not using active directory features inside lambda ([Ref](https://github.com/aws/amazon-redshift-python-driver/blob/2bcbaf2ffdc1edbb7260eab2789f640dc579f590/redshift_connector/plugin/adfs_credentials_provider.py#L66))
   -  Occupies ~10 MB space
   - Has native C libraries packaged with it and this is platform dependent and a direct zip and upload wont work (i am using a macosx for development and lambda uses amazon linux)

I also see pytz, however as of now i am not sure of the impact of removing the same

**Comment by Brooke-white** [comment, marked as answer]
Thank you for this feedback. Regarding the dependency on lxml, this package is uses for some of our IdP (Identity provider plugin) code along with beautifulsoup. In looking at the beautifulsoup docs I see there is a HTML parser by default, from Python standard library [ref](https://www.crummy.com/software/BeautifulSoup/bs4/doc/#installing-a-parser).

I think it'd be worth while to investigate if the parser included by default would be sufficient here (we haven't tested this), as it'd be nice to reduce the number of dependencies redshift_connector has. I will take a look into this option and follow up with my findings. 

While lxml is not applicable to your use case, there are methods for downloading Python modules for a platform other than the one currently in use. [see this stackoverflow post](https://stackoverflow.com/questions/59083607/how-to-download-pip-module-for-linux-using-windows) for example. I had no clue this was an option, but it's pretty cool :)

As for the dependencies on boto3 and botocore, these are needed for use of IAM and IDP auth. Though needing to remove them before deploying your lambda is not ideal as they take up a good bit of space... ~67MB or so. Aside from taking up space, the lambda docs imply this shouldn't break anything:

> Note: Lambda periodically updates the Boto3 libraries to enable the latest set of features and security updates. To have full control of the dependencies your function uses, package all of your dependencies with your deployment package. [ref](https://docs.aws.amazon.com/lambda/latest/dg/python-package.html)


As for documentation re: using redshift_connector with lambda, I've used the official AWS docs for [Deploy Python Lambda functions with .zip file archives](https://docs.aws.amazon.com/lambda/latest/dg/python-package.html) in the past.

Have you found gaps specific to redshift_connector in these docs that you'd like to see addressed in our docs? If so, please let us know and we can add some documentation :)
